File: year2015/day05/python/solution.py
import re

def main():    
    print(part_1())
    print(part_2())


def part_1():
    return validate([r'^(?!.*ab)(?!.*cd)(?!.*pq)(?!.*xy).*$', r'^(.*[aeiou]){3}.*$', r'(.)\1'], input)


def part_2():
    return validate([r'([a-z][a-z]).*\1', r'([a-z])[a-z]\1'], input)


def setup(path):
    global input
    
    with open(path, 'r') as file:
        input = [row for row in file]


def validate(valid, input):
    counter = 0
    for string in input:
        if all(re.search(pattern, string) for pattern in valid):
            counter += 1

    return counter


if __name__ == '__main__':
    setup('../puzzle.txt')
    main()

File: year2015/day05/java/Solution.java
import java.nio.file.*;
import java.util.*;
import java.util.regex.*;

public class Solution {
    private static List<String> STRINGS;

    public static void main(String[] args) { 
        try {
            setup(Files.readAllLines(Paths.get("../puzzle.txt")));
            System.out.println(part1());
            System.out.println(part2());
        } catch(Exception e) {
            e.printStackTrace();
        }
    }
    
	public static int part1() {
        return validate(List.of("^(?!.*ab)(?!.*cd)(?!.*pq)(?!.*xy).*$", "^(.*[aeiou]){3}.*$", "(.)\\1"));
    }
    
    public static int part2() {
        return validate(List.of("([a-z][a-z]).*\\1", "([a-z])[a-z]\\1"));
    }

    public static void setup(List<String> input) {
        STRINGS = input;
    }

    private static int validate(List<String> validators) {
        return (int) STRINGS.stream().filter(str -> validators.stream().allMatch(pattern -> Pattern.compile(pattern).matcher(str).find())).count();
    }
}